lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+		defer stopWg.Done()
+		errs[1] = fh.votingRound.Stop()
+	}()
+
+	stopWg.Wait()
+
+	if errs[0] != nil {
+		return errs[0]
+	}
+
+	if errs[1] != nil {
+		return errs[1]
+	}
+
+	return nil
+}
+
+func (fh *finalizationHandler) Stop() (err error) {
+	close(fh.stopCh)
+
+	err = fh.stop()
+	select {
+	case <-fh.handlerDone:
+	case <-time.After(fh.timeoutStop):
+		return errTimeoutWhileStoping
+	}
+
+	close(fh.observableErrs)
+	return err
+}
+
+// waitServices will start the services and wait until they complete or
+func (fh *finalizationHandler) waitServices() error {
+	fh.servicesLock.Lock()
+	fh.finalizationEngine, fh.votingRound = fh.newServices()
+	fh.servicesLock.Unlock()
+
+	finalizationEngineErr := make(chan error)
+	go func() {
+		defer close(finalizationEngineErr)
+		err := fh.finalizationEngine.Start()
+		if err != nil {
+			finalizationEngineErr <- err
+		}
+	}()
+
+	votingRoundErr := make(chan error)
+	go func() {
+		defer close(votingRoundErr)
+		err := fh.votingRound.Start()
+		if err != nil {
+			votingRoundErr <- err
+		}
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return nil
+
+		case err, ok := <-votingRoundErr:
+			if !ok {
+				votingRoundErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+			return err
+
+		case err, ok := <-finalizationEngineErr:
+			if !ok {
+				finalizationEngineErr = nil
+			}
+
+			stopErr := fh.stop()
+			if stopErr != nil {
+				logger.Infof("stopping finalisation handler: %s", stopErr)
+			}
+
+			return err
+
+		default:
+			finish := votingRoundErr == nil && finalizationEngineErr == nil
+			if finish {
+				return nil
+			}
+		}
+	}
+}
+
+var errTimeoutWhileStoping = errors.New("timeout while stopping")
+
+type handleVotingRound struct {
+	grandpaService       *Service
+	timeoutStop          time.Duration
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		timeoutStop:          5 * time.Second,
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.stopCh == nil {
+		return nil
+	}
+
+	close(h.stopCh)
+	select {
+	case <-h.engineDone:
+	case <-time.After(h.timeoutStop):
+		return errTimeoutWhileStoping
+	}
+
+	h.stopCh = nil
+	return nil
+}
+
+// playGrandpaRound executes a round of GRANDPA
+// at the end of this round, a block will be finalised.
+// TODO(test): stopping handleVotingRound first and then stopping
+// finalizationEngine might cause a write in a non-reading unbuff channel
+// blocking the finalizationEngine to stop and triggering the stop timeout timer
+func (h *handleVotingRound) Start() (err error) {
```

Comment:
This is not really a start function since it runs until stopped. Usually/imo/afaik, Start should launch a goroutine and return a start error promptly. The easiest fix here is to rename `Start` to `Run`

Reply:
I agree with you, I changed the `ephemeralService` interface to be `Run(), Stop()` and now makes more sense to understand that we **run** an ephemeral service but we allow the caller to **stop** it